### ISSUE TITLE
Slack backend gets WebSocketConnectionClosedException then no longer responds to messages

### ISSUE THREAD

**Issue by mattcl** [skoczen/will#331]
I'm guessing that after this exception is raised either the process stops listening/doesn't reconnect or something gets stuck. When this occurs my only option seems to be restarting the bot. I'm running at a DEBUG logging level, but this is the only output that that wasn't about starting new http connections.
```
Fri, 19 Jan 2018 06:50:50 [CRITICAL] Error in watching slack RTM: 
Traceback (most recent call last):
  File "/usr/local/lib/python3.4/dist-packages/will/backends/io_adapters/slack.py", line 424, in _watch_slack_rtm
    events = self.client.rtm_read()
  File "/usr/local/lib/python3.4/dist-packages/slackclient/client.py", line 128, in rtm_read
    json_data = self.server.websocket_safe_read()
  File "/usr/local/lib/python3.4/dist-packages/slackclient/server.py", line 186, in websocket_safe_read
    data += "{0}\n".format(self.websocket.recv())
  File "/usr/local/lib/python3.4/dist-packages/websocket/_core.py", line 296, in recv
    opcode, data = self.recv_data()
  File "/usr/local/lib/python3.4/dist-packages/websocket/_core.py", line 313, in recv_data
    opcode, frame = self.recv_data_frame(control_frame)
  File "/usr/local/lib/python3.4/dist-packages/websocket/_core.py", line 326, in recv_data_frame
    frame = self.recv_frame()
  File "/usr/local/lib/python3.4/dist-packages/websocket/_core.py", line 360, in recv_frame
    return self.frame_buffer.recv_frame()
  File "/usr/local/lib/python3.4/dist-packages/websocket/_abnf.py", line 340, in recv_frame
    self.recv_header()
  File "/usr/local/lib/python3.4/dist-packages/websocket/_abnf.py", line 288, in recv_header
    header = self.recv_strict(2)
  File "/usr/local/lib/python3.4/dist-packages/websocket/_abnf.py", line 375, in recv_strict
    bytes_ = self.recv(min(16384, shortage))
  File "/usr/local/lib/python3.4/dist-packages/websocket/_core.py", line 430, in _recv
    return recv(self.sock, bufsize)
  File "/usr/local/lib/python3.4/dist-packages/websocket/_socket.py", line 94, in recv
    "Connection is already closed.")
websocket._exceptions.WebSocketConnectionClosedException: Connection is already closed.
```

**Comment by netjunki** [comment]
Yeah I've see this as well. It seems like this happens when slack has an outage or any sort of network hiccup happens between you and slack (like IT restarts the switch your server is connected to).

**Comment by netjunki** [comment]
Been seeing a similar issue with a stack trace like this:

    Traceback (most recent call last):
      File "/home/benlau/code/slackbot/will/will/backends/io_adapters/slack.py", line 451, in _watch_slack_rtm
        events = self.client.rtm_read()
      File "/home/benlau/.local/lib/python2.7/site-packages/slackclient/_client.py", line 123, in rtm_read
        json_data = self.server.websocket_safe_read()
      File "/home/benlau/.local/lib/python2.7/site-packages/slackclient/_server.py", line 141, in websocket_safe_read
        data += "{0}\n".format(self.websocket.recv())
      File "/home/benlau/.local/lib/python2.7/site-packages/websocket/_core.py", line 293, in recv
        opcode, data = self.recv_data()
      File "/home/benlau/.local/lib/python2.7/site-packages/websocket/_core.py", line 310, in recv_data
        opcode, frame = self.recv_data_frame(control_frame)
      File "/home/benlau/.local/lib/python2.7/site-packages/websocket/_core.py", line 323, in recv_data_frame
        frame = self.recv_frame()
      File "/home/benlau/.local/lib/python2.7/site-packages/websocket/_core.py", line 357, in recv_frame
        return self.frame_buffer.recv_frame()
      File "/home/benlau/.local/lib/python2.7/site-packages/websocket/_abnf.py", line 336, in recv_frame
        self.recv_header()
      File "/home/benlau/.local/lib/python2.7/site-packages/websocket/_abnf.py", line 286, in recv_header
        header = self.recv_strict(2)
      File "/home/benlau/.local/lib/python2.7/site-packages/websocket/_abnf.py", line 371, in recv_strict
        bytes_ = self.recv(min(16384, shortage))
      File "/home/benlau/.local/lib/python2.7/site-packages/websocket/_core.py", line 427, in _recv
        return recv(self.sock, bufsize)
      File "/home/benlau/.local/lib/python2.7/site-packages/websocket/_socket.py", line 80, in recv
        bytes_ = sock.recv(bufsize)
      File "/usr/lib/python2.7/ssl.py", line 341, in recv
        return self.read(buflen)
      File "/usr/lib/python2.7/ssl.py", line 260, in read
        return self._sslobj.read(len)
   error: [Errno 110] Connection timed out

**Comment by cmachine** [comment]
I'm also seeing this error. Didn't realize there was an open issue/PR for it and just put in my own too: #334

**Comment by skoczen** [comment]
(Quick paste-on-all-open-issues update:)

**TL;DR**: New Will release EOM. Also, you're awesome.

**Full Update:** This may be TMI for you, and if so, cool, the thing you probably care about is above.  But given the total radio silence on my part - and the fact that I've always tried to create a community here where we got to be actual people instead of avatars and PRs, I wanted to give a bit more context.

As you may be aware from an earlier AFK, I got married in December, which was great, but burned the month from a Will perspective.  I planned to get back on it at the start of January, but instead found myself selling our house, and simultaneously hit with a surprisingly severe bout of depression.  The time since Jan 1 has been sort of a blur.  A dark blur, but a blur more than anything.   I saw the PRs and comments coming in, but didn't have the energy or time to do anything about them.

Things seem to be clearing up on both counts now, and I'm _really_ looking forward to getting these fixes in, catching up on bug reports, and being active in improving Will again. :)

With that, I'm aiming to get all outstanding PRs bundled into a release, and all issues replied to by the end of February.

Most of all, thanks for your contribution to Will, your patience with an imperfect, quite human project maintainer, and for continuing to come back, engage, and spend some of the precious seconds you get on earth making this project better.

I appreciate you.  Thank you.

**Comment by skoczen** [comment]
This has been fixed, and should now work in 2.1.1!
